Mathlib/Algebra/GroupPower/Lemmas.lean

```diff
@@ -2,35 +2,1316 @@
+  induction' k with k ih <;> [rfl, rw [pow_succ', Int.natAbs_mul, pow_succ', ih]]
+#align int.nat_abs_pow Int.nat_abs_pow
+
+section bit0_bit1
+
+set_option linter.deprecated false
+
+-- The next four lemmas allow us to replace multiplication by a numeral with a `zsmul` expression.
+-- They are used by the `noncomm_ring` tactic, to normalise expressions before passing to `abel`.
+theorem bit0_mul [NonUnitalNonAssocRing R] {n r : R} : bit0 n * r = (2 : ℤ) • (n * r) := by
+  dsimp [bit0]
+  rw [add_mul, ← one_add_one_eq_two, add_zsmul, one_zsmul]
+#align bit0_mul bit0_mul
+
+theorem mul_bit0 [NonUnitalNonAssocRing R] {n r : R} : r * bit0 n = (2 : ℤ) • (r * n) := by
+  dsimp [bit0]
+  rw [mul_add, ← one_add_one_eq_two, add_zsmul, one_zsmul]
+#align mul_bit0 mul_bit0
+
+theorem bit1_mul [NonAssocRing R] {n r : R} : bit1 n * r = (2 : ℤ) • (n * r) + r := by
+  dsimp [bit1]
+  rw [add_mul, bit0_mul, one_mul]
+#align bit1_mul bit1_mul
+
+theorem mul_bit1 [NonAssocRing R] {n r : R} : r * bit1 n = (2 : ℤ) • (r * n) + r := by
+  dsimp [bit1]
+  rw [mul_add, mul_bit0, mul_one]
+#align mul_bit1 mul_bit1
+
+end bit0_bit1
+
+@[simp]
+theorem zsmul_eq_mul [Ring R] (a : R) : ∀ n : ℤ, n • a = n * a
+  | (n : ℕ) => by rw [coe_nat_zsmul, nsmul_eq_mul, Int.cast_ofNat]
+  | -[n+1] => by simp [Nat.cast_succ, neg_add_rev, Int.cast_negSucc, add_mul]
+#align zsmul_eq_mul zsmul_eq_mul
+
+theorem zsmul_eq_mul' [Ring R] (a : R) (n : ℤ) : n • a = a * n := by
+  rw [zsmul_eq_mul, (n.cast_commute a).eq]
+#align zsmul_eq_mul' zsmul_eq_mul'
+
+/-- Note that `add_comm_group.int_smul_comm_class` requires stronger assumptions on `R`. -/
+instance NonUnitalNonAssocRing.int_smul_comm_class [NonUnitalNonAssocRing R] :
+    SMulCommClass ℤ R R :=
+  ⟨fun n x y =>
+    match n with
+    | (n : ℕ) => by simp_rw [coe_nat_zsmul, smul_comm]
+    | -[n+1] => by simp_rw [negSucc_zsmul, smul_eq_mul, mul_neg, mul_smul_comm]⟩
+#align non_unital_non_assoc_ring.int_smul_comm_class NonUnitalNonAssocRing.int_smul_comm_class
+
+/-- Note that `add_comm_group.int_is_scalar_tower` requires stronger assumptions on `R`. -/
+instance NonUnitalNonAssocRing.int_is_scalar_tower [NonUnitalNonAssocRing R] :
+    IsScalarTower ℤ R R :=
+  ⟨fun n x y =>
+    match n with
+    | (n : ℕ) => by simp_rw [coe_nat_zsmul, smul_assoc]
+    | -[n+1] => by simp_rw [negSucc_zsmul, smul_eq_mul, neg_mul, smul_mul_assoc]⟩
+#align non_unital_non_assoc_ring.int_is_scalar_tower NonUnitalNonAssocRing.int_is_scalar_tower
+
+theorem zsmul_int_int (a b : ℤ) : a • b = a * b := by simp
+#align zsmul_int_int zsmul_int_int
+
+theorem zsmul_int_one (n : ℤ) : n • (1 : ℤ) = n := by simp
+#align zsmul_int_one zsmul_int_one
+
+@[simp, norm_cast]
+theorem Int.cast_pow [Ring R] (n : ℤ) (m : ℕ) : (↑(n ^ m) : R) = (↑n : R) ^ m := by
+  induction' m with m ih
+  · rw [pow_zero, pow_zero, Int.cast_one]
+  · rw [pow_succ, pow_succ, Int.cast_mul, ih]
+#align int.cast_pow Int.cast_powₓ
```

Comment:
Do you really mean to leave the little x? My guess/hope is that whatever the difference is, it doesn't matter, because the declarations are "morally the same". Add "porting note"? It looks like the problem is simply that casts changed but I don't think other lemmas which use this are going to care.

Reply:
This was Scott Morrison's advice in the cases where the alignment was not exact.